### PR TITLE
Feature/relative days

### PR DESCRIPTION
**Pull request by MatthiasGu**

### CODE REVIEW

File: src/filters/dates/relative/relativeDates.ts [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+            quantity = 1;
+        } else if (contains(period, RELATIVE_DATES.VERBAL_QUANTIFIERS.NEXT)) {
+            return this.getNext(unit);
+        } else {
+            quantity = parseInt(period, 10);
+        }
+        if (contains(unit, RELATIVE_DATES.TIME_UNITS.ANY)) {
+            return DateTime.utc().plus({ [unit]: quantity });
+        }
+        return null;
```
Comment (MatthiasGu) [inline src/filters/dates/relative/relativeDates.ts]: Not sure how to handle this and the other return nulls inside helper methods, at the moment they will never return null, because the if conditions are exhaustive, i.e. the function is only called if the pattern is matched, and the function goes through all the possible pattern matches under current rules. It is necessary to have the return null though due to methods having to always return the given type (and the compiler does not know that all possibilities are exhausted). The workaround would be to have the current last condition as default return (in this case line 58 would not have the if condition. However, this IMO reduces the clarify and might lead to bugs in the future when we add more possibilities. Thoughts?

File: src/filters/dates/relative/relativeDates.constants.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+const TOMORROW = 'tomorrow';
+
+const RELATIVE_ADVERB = {
+    ANY: `${TODAY}|${TOMORROW}`,
+    TODAY: `${TODAY}`,
+    TOMORROW: `${TOMORROW}`,
+};
+
+const RELATIVE_PREPOSITIONS = '((in|after) )?';
+const RELATIVE_POSTPOSTIONS = ' (from now|later|after)';
```
Comment (h-dong) [inline src/filters/dates/relative/relativeDates.constants.ts]: Can we write it like this so it is clearer

```suggestion
const RELATIVE_POSTPOSTIONS = '( (from now)|(later)|(after))';
```

File: src/filters/dates/relative/relativeDates.test.ts [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+            ],
+        }, {
+            description: 'should parse in a yr correctly',
+            case: 'go to work in a yr',
+            result: [
+                { dateTime: mockDates(currentDay, currentMonth, currentYear + 1), text: 'go to work', matched: 'in a yr' },
+            ],
+        }, {
+            description: 'should parse in X yrs correctly',
+            case: 'go to work in 5 yrs',
```
Comment (h-dong) [inline src/filters/dates/relative/relativeDates.test.ts:146]: You are going to say scope creep again, what about `go to work in five yrs`?

File: src/filters/dates/relative/relativeDates.test.ts [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+
+        testCases.forEach(item => {
+            test(item.description, () => {
+                const parsedText = RelativeDates.parseText(item.case);
+                expect(parsedText).toEqual(item.result);
+            });
+        });
+    });
+
+    describe('getNext', () => {
```
Comment (h-dong) [inline src/filters/dates/relative/relativeDates.test.ts:182]: I don't think these tests make sense, we are essentially test the `DateTime` library.

Should test by mocking e.g.

```
    static getNext(unit): DateTime {
        return DateTime.utc()
            .plus({ [unit]: 1 })
            .startOf(unit)
            .endOf('day');
    }


    test('getNext()', () => {
        const spy = jest.spyOn(DateTime, 'utc');
         // blah blah blah
        expect(spy).toHaveBeenCalledWith(blah);
    });
```

Comment (h-dong) [reply src/filters/dates/relative/relativeDates.test.ts:182]: Also why is this a static method? It should be utils function. Or even just stick it outside of the class.

File: src/filters/dates/relative/relativeDates.ts [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        if (!matches) return null;
+
+        return matches.map(match => {
+            const dateTimeObj = this.convertMatchToDateObj(match);
+            return parseMatches(text, match, dateTimeObj);
+        });
+    }
+
+    static timeUnitToString(matchAgainst: string): string {
+        for (const key in RELATIVE_DATES.TIME_UNITS) {
```
Comment (h-dong) [inline src/filters/dates/relative/relativeDates.ts]: Remember how we said we are not going to use `for in` loops lol

File: src/filters/dates/relative/relativeDates.ts (hunk 1 above)
Comment (h-dong) [reply src/filters/dates/relative/relativeDates.ts]: Have no idea what you are trying to say here lol

**Review by h-dong — COMMENTED** [review]
You are also missed 2 lines in test coverage report

**Comment by h-dong** [comment]
Why is `got to work in 1000` matching? Did we cover this already in the other PR?

**Comment by h-dong** [comment]
Probably edge cases `got to work in 10000000 days` (return "dateTime": "+029398-10-19T23:59:59.999Z",` and `got to work in 100000000 days` (return "dateTime": null)

**Comment by h-dong** [comment]
Maybe should match `2 weeks from today go to work` too?

File: src/filters/dates/relative/relativeDates.test.ts [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+
+// Mock Date Time to Thu Jun 20 2019 08:34:52 GMT+0100
+Settings.now = () => 1561019692628;
+const currentYear = DateTime.utc().year;
+const currentMonth = DateTime.utc().month;
+const currentDay = DateTime.utc().day;
+
+beforeEach(() => {
+    // Restore mocked time in case we need to change it for individual tests
+    Settings.now = () => 1561019692628;
```
Comment (h-dong) [inline src/filters/dates/relative/relativeDates.test.ts]: Add a human readable comment here, what date is this?

File: src/filters/dates/relative/relativeDates.test.ts [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+import RelativeDates from './relativeDates';
+
+// Mock Date Time to Thu Jun 20 2019 08:34:52 GMT+0100
+Settings.now = () => 1561019692628;
+const currentYear = DateTime.utc().year;
+const currentMonth = DateTime.utc().month;
+const currentDay = DateTime.utc().day;
+
+beforeEach(() => {
+    // Restore mocked time in case we need to change it for individual tests
```
Comment (h-dong) [inline src/filters/dates/relative/relativeDates.test.ts]: We are not restoring mock, we are mocking.

File: src/filters/dates/relative/relativeDates.test.ts [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+    Settings.now = () => Date.now();
+});
+
+describe('RelativeDates', () => {
+    const mockDates = (day, month, year) => DateTime.utc()
+        .set({ day, month, year })
+        .endOf('day')
+        .toJSDate();
+
+    const mockNext = (period) => DateTime.utc()
```
Comment (h-dong) [inline src/filters/dates/relative/relativeDates.test.ts:28]: The mock is basically the original function 😂

File: src/filters/dates/relative/relativeDates.test.ts [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+                { dateTime: mockDates(currentDay, currentMonth, currentYear + 1), text: 'go to work', matched: 'in a year' },
+            ],
+        }, {
+            description: 'should parse in X years correctly',
+            case: 'go to work in 5 years',
+            result: [
+                { dateTime: mockDates(currentDay, currentMonth, currentYear + 5), text: 'go to work', matched: 'in 5 years' },
+            ],
+        }, {
+            description: 'should parse next year correctly',
```
Comment (h-dong) [inline src/filters/dates/relative/relativeDates.test.ts:133]: `go to work last year` doesn't match "last year", is this expected?

File: src/filters/dates/relative/relativeDates.constants.ts (hunk 2 above)
Comment (MatthiasGu) [reply src/filters/dates/relative/relativeDates.constants.ts]: What do you mean by "wrong"? It is passing tests.

File: src/filters/dates/relative/relativeDates.test.ts (hunk 3 above)
Comment (MatthiasGu) [reply src/filters/dates/relative/relativeDates.test.ts:146]: Are you going to write a regex that will accept any verbal figure? That could be difficult.

File: src/filters/dates/relative/relativeDates.test.ts (hunk 4 above)
Comment (MatthiasGu) [reply src/filters/dates/relative/relativeDates.test.ts:182]: I think they make sense, we are testing that the correct unit is being passed to DateTime. Regardless, if the library that we rely on fails, we ought to know that as we expose that to the user.

Comment (MatthiasGu) [reply src/filters/dates/relative/relativeDates.test.ts:182]: > Also why is this a static method? It should be utils function. Or even just stick it outside of the class.

I would prefer to do this as part of a separate PR whereby I will refactor helper methods into proper folder structure.

File: src/filters/dates/relative/relativeDates.ts (hunk 5 above)
Comment (MatthiasGu) [reply src/filters/dates/relative/relativeDates.ts]: What is the problem with it? We need to loop over object properties and that's what it does.

**Comment by MatthiasGu** [comment]
> Why is `got to work in 1000` matching? Did we cover this already in the other PR?

Do you know which parser is matching this? I added a couple of test cases, it does not seem to match the most recent parses.

**Comment by MatthiasGu** [comment]
> Probably edge cases `got to work in 10000000 days` (return "dateTime": "+029398-10-19T23:59:59.999Z",`and`got to work in 100000000 days` (return "dateTime": null)

Probably to do with limitations of DateTime, I don't think we need to care about this.

**Comment by MatthiasGu** [comment]
> Maybe should match `2 weeks from today go to work` too?

Are you referring to the different order? That works. If you mean "from today", it is always the matter of simply adding more keywords to the list and then it will match.

File: src/filters/dates/relative/relativeDates.ts (hunk 1 above)
Comment (MatthiasGu) [reply src/filters/dates/relative/relativeDates.ts]: Codecov is complaining that this null check is never hit. The reason it is never hit is that at the moment, the above condition is always true (otherwise this function would not be called, as time units are mandatory for this expression to match).

However, the null check is necessary, because the compiler does not know that the if condition is always met and complains if there is no default return.

File: src/filters/dates/relative/relativeDates.constants.ts (hunk 2 above)
Comment (h-dong) [reply src/filters/dates/relative/relativeDates.constants.ts]: Initially I thought it was wrong, then I realised I misread the brackets. But I didn't delete the original message properly it seems

File: src/filters/dates/relative/relativeDates.ts (hunk 5 above)
Comment (h-dong) [reply src/filters/dates/relative/relativeDates.ts]: Not ES5, that's all

File: src/filters/dates/relative/relativeDates.test.ts (hunk 4 above)
Comment (h-dong) [reply src/filters/dates/relative/relativeDates.test.ts:182]: > I think they make sense, we are testing that the correct unit is being passed to DateTime. Regardless, if the library that we rely on fails, we ought to know that as we expose that to the user.

Yeah it makes sense to test, but it doesn't make sense to test the library. If you use lodash you wouldn't test the functions of lodash right?

File: src/filters/dates/relative/relativeDates.test.ts (hunk 3 above)
Comment (h-dong) [reply src/filters/dates/relative/relativeDates.test.ts:146]: It is weird to allow "a" and "1" but not "one", don't worry about it for now I'll add an issue for this

File: src/filters/dates/relative/relativeDates.test.ts (hunk 4 above)
Comment (h-dong) [reply src/filters/dates/relative/relativeDates.test.ts:182]: We can leave this for now, just not a good practice imo

**Comment by h-dong** [comment]
> > Why is `got to work in 1000` matching? Did we cover this already in the other PR?
> 
> Do you know which parser is matching this? I added a couple of test cases, it does not seem to match the most recent parses.

Try it out on deploy/netlify build, I'm not sure which one is matching that.

**Comment by h-dong** [comment]
> > Maybe should match `2 weeks from today go to work` too?
> 
> Are you referring to the different order? That works. If you mean "from today", it is always the matter of simply adding more keywords to the list and then it will match.

It works for "2 weeks from now", but doesn't for "2 weeks from today"?

File: src/filters/dates/relative/relativeDates.constants.ts (hunk 2 above)
Comment (h-dong) [reply src/filters/dates/relative/relativeDates.constants.ts]: I think for this comment you literally just need to click on accept. Don't need code change

File: src/filters/dates/relative/relativeDates.ts (hunk 5 above)
Comment (h-dong) [reply src/filters/dates/relative/relativeDates.ts]: I'll go through it all and change it later

File: src/filters/dates/relative/relativeDates.constants.ts (hunk 2 above)
Comment (MatthiasGu) [reply src/filters/dates/relative/relativeDates.constants.ts]: I don't think the extra brackets are needed though, does it really add clarity?

Comment (h-dong) [reply src/filters/dates/relative/relativeDates.constants.ts]: Yeah the space is more obvious now

File: src/filters/dates/relative/relativeDates.test.ts (hunk 9 above)
Comment (MatthiasGu) [reply src/filters/dates/relative/relativeDates.test.ts:133]: Yes, since we are only doing future words at the moment, we can always add past words if a use case arises. We don't have stuff like "ago" either.

File: src/filters/dates/relative/relativeDates.test.ts (hunk 6 above)
Comment (MatthiasGu) [reply src/filters/dates/relative/relativeDates.test.ts]: > > > Maybe should match `2 weeks from today go to work` too?
> > 
> > 
> > Are you referring to the different order? That works. If you mean "from today", it is always the matter of simply adding more keywords to the list and then it will match.
> 
> It works for "2 weeks from now", but doesn't for "2 weeks from today"?

Because "from today" wasn't in the list of expected expressions, but I added it now. In the future, if you want to add one you can in relativeDates.constants.ts

**Comment by MatthiasGu** [comment]
Found a way to simplify the logic, codecov does not complain anymore.

**Review by h-dong — APPROVED** [review]
